cli/cli_app.go

```diff
@@ -548,10 +554,31 @@ func runTerragruntWithConfig(terragruntOptions *options.TerragruntOptions, terra
 	}
 
 	return runActionWithHooks("terraform", terragruntOptions, terragruntConfig, func() error {
-		return runTerraformWithRetry(terragruntOptions)
+		runTerraformError := runTerraformWithRetry(terragruntOptions)
+
+		var lockFileError error
+		if util.FirstArg(terragruntOptions.TerraformCliArgs) == CMD_INIT {
+			lockFileError = copyLockFile(terragruntOptions.WorkingDir, originalTerragruntOptions.WorkingDir, terragruntOptions.Logger)
```

Comment:
This could use a comment because it took me a bit to follow what's going on, or improvement.

Am I correct in understanding that what we are trying to accomplish here is to get the lock file to be copied to where the terragrunt config is? And this works because of https://github.com/gruntwork-io/terragrunt/blob/master/options/options.go#L207 ?

If that's correct, I feel like this makes more sense if we use the directory where `terragruntOptions.TerragruntConfigPath` is. I think the direct implementation would make this easier to follow in terms of what we are trying to accomplish here, and we can avoid passing around the `originalTerragruntOptions`.

Reply:
Hm, I could've sworn that the "working dir" could be different in the past from the dir where the `terragrunt.hcl` file was... E.g., You could run `cd /foo && terragrunt apply --terragrunt-config /bar/terragrunt.hcl` and  working dir used to be set to `/foo` even though `terragrunt.hcl` was in `/bar`... But looking at `terragrunt.Options`, at least, that doesn't seem to be the case. Am I just mis-remembering? Or did that change at some point?

Reply:
> But looking at terragrunt.Options, at least, that doesn't seem to be the case. Am I just mis-remembering? Or did that change at some point?

Honestly speaking, I am not sure. This is one of those areas that is extremely difficult to follow in the terragrunt codebase, especially when you take into account `dependency`, `xxx-all`, etc.

So all the more reason that we either:

- Need a comment here with the true intent of where we want the lock file to eventually be.
- Update the code, if possible, to be more clear/direct about where we place the lock file.

> You could run cd /foo && terragrunt apply --terragrunt-config /bar/terragrunt.hcl and working dir used to be set to /foo even though terragrunt.hcl was in /bar...

Even if that were the case, is that where we want the lock file? It seems like we want it where the terragrunt config is, right?

Reply:
> Even if that were the case, is that where we want the lock file (in `/foo`)? It seems like we want it where the terragrunt config is (in `/bar`), right? Since we only copy things out of where the `terragrunt.hcl` file is located?

You know, I'm not 100% sure. I could see arguments for either one. I also realize we need to document the behavior, not just in a comment as you said, but in the official docs. I won't have time to get to this today, so I'll have to give this more thought and come back and wrap it up later. Thx for the great questions.

Reply:
Alright, added a comment explaining this behavior in https://github.com/gruntwork-io/terragrunt/pull/1459/commits/31830b5aa13806267bea785d157290914fd105f7 and a whole new page in the documentation that describes how Terragrunt handles lock files in https://github.com/gruntwork-io/terragrunt/pull/1459/commits/c15eefbb1dc8eebf336c334e9d78f011c4798ff9.